### ISSUE TITLE
Cannot convert time to DateTime

### ISSUE THREAD

**Issue by Balinus** [Alexander-Barth/NCDatasets.jl#12]
Hello!

I was playing with the time vector when I realized that your package converts to a DateTime type. Hence, I'm thinking that I could simply leverage your package for this task. 

However, using multiples CORDEX and CMIP5 nc files, I can't convert the time variable to DateTime and get errors. I've added one of the file that are causing problem. I can provide more files. Thanks for any hints! :+1: 

https://www.dropbox.com/sh/1gjfdjpb0rtwq2w/AAAn19nXqu5wgeXGoxY97MS2a?dl=0

```julia
ds=Dataset(cmip5file2)
Dataset: /path/to/file/tasmax_day_IPSL-CM5A-MR_historical_r2i1p1_20000101-20051231.nc
Group: /

Dimensions
   time = 2190
   lat = 143
   lon = 144
   bnds = 2

Variables
  time   (2190)
    Datatype:    Float64
    Dimensions:  time
    Attributes:
     bounds               = time_bnds
     units                = days since 1850-01-01 00:00:00
     calendar             = noleap
     axis                 = T
     long_name            = time
     standard_name        = time

  time_bnds   (2 × 2190)
    Datatype:    Float64
    Dimensions:  bnds × time

  lat   (143)
    Datatype:    Float64
    Dimensions:  lat
    Attributes:
     bounds               = lat_bnds
     units                = degrees_north
     axis                 = Y
     long_name            = latitude
     standard_name        = latitude

  lat_bnds   (2 × 143)
    Datatype:    Float64
    Dimensions:  bnds × lat

  lon   (144)
    Datatype:    Float64
    Dimensions:  lon
    Attributes:
     bounds               = lon_bnds
     units                = degrees_east
     axis                 = X
     long_name            = longitude
     standard_name        = longitude

  lon_bnds   (2 × 144)
    Datatype:    Float64
    Dimensions:  bnds × lon

  height  
    Attributes:
     units                = m
     axis                 = Z
     positive             = up
     long_name            = height
     standard_name        = height

  tasmax   (144 × 143 × 2190)
    Datatype:    Float32
    Dimensions:  lon × lat × time
    Attributes:
     standard_name        = air_temperature
     long_name            = Daily Maximum Near-Surface Air Temperature
     units                = K
     original_name        = t2m_max
     cell_methods         = time: maximum (interval: 30 minutes)
     cell_measures        = area: areacella
     history              = 2012-04-25T17:00:57Z altered by CMOR: Treated scalar dimension: 'height'. 2012-04-25T17:00:57Z altered by CMOR: replaced missing value flag (9.96921e+36) with standard missing value (1e+20). 2012-04-25T17:00:58Z altered by CMOR: Inverted axis: lat.
     coordinates          = height
     missing_value        = 1.0e20
     _FillValue           = 1.0e20
     associated_files     = baseURL: http://cmip-pcmdi.llnl.gov/CMIP5/dataLocation gridspecFile: gridspec_atmos_fx_IPSL-CM5A-MR_historical_r0i0p0.nc areacella: areacella_fx_IPSL-CM5A-MR_historical_r0i0p0.nc

Global attributes
  institution          = IPSL (Institut Pierre Simon Laplace, Paris, France)
  institute_id         = IPSL
  experiment_id        = historical
  source               = IPSL-CM5A-MR (2010) : atmos : LMDZ4 (LMDZ4_v5, 144x143x39); ocean : ORCA2 (NEMOV2_3, 2x2L31); seaIce : LIM2 (NEMOV2_3); ocnBgchem : PISCES (NEMOV2_3); land : ORCHIDEE (orchidee_1_9_4_AR5)
  model_id             = IPSL-CM5A-MR
  forcing              = Nat,Ant,GHG,SA,Oz,LU,SS,Ds,BC,MD,OC,AA
  parent_experiment_id = piControl
  parent_experiment_rip = r1i1p1
  branch_time          = 1860.0
  contact              = ipsl-cmip5 _at_ ipsl.jussieu.fr Data manager : Sebastien Denvil
  comment              = This 20th century simulation include natural and anthropogenic forcings.
  references           = Model documentation and further reference available here : http://icmc.ipsl.fr
  initialization_method = 1
  physics_version      = 1
  tracking_id          = 865ced25-4e24-43d1-bf07-fd163ba5c7b7
  product              = output
  experiment           = historical
  frequency            = day
  creation_date        = 2012-04-25T17:00:58Z
  history              = 2012-04-25T17:00:58Z CMOR rewrote data to comply with CF standards and CMIP5 requirements.
  Conventions          = CF-1.4
  project_id           = CMIP5
  table_id             = Table day (10 February 2011) 80e409bd73611e9d25d049ad2059c310
  title                = IPSL-CM5A-MR model output prepared for CMIP5 historical
  parent_experiment    = pre-industrial control
  modeling_realm       = atmos
  realization          = 2
  cmor_version         = 2.7.1


julia> ds["time"]
time (2190)
  Datatype:    Float64
  Dimensions:  time
  Attributes:
   bounds               = time_bnds
   units                = days since 1850-01-01 00:00:00
   calendar             = noleap
   axis                 = T
   long_name            = time
   standard_name        = time

julia> ds["time"][:]
ERROR: MethodError: no method matching start(::Void)
Closest candidates are:
  start(::SimpleVector) at essentials.jl:258
  start(::Base.MethodList) at reflection.jl:560
  start(::ExponentialBackOff) at error.jl:107
  ...
Stacktrace:
 [1] timedecode(::Array{Float64,1}, ::String, ::String) at /home/proy/.julia/v0.6/NCDatasets/src/NCDatasets.jl:217
 [2] getindex(::NCDatasets.CFVariable{Float64,Float64,1}, ::Colon) at /home/proy/.julia/v0.6/NCDatasets/src/NCDatasets.jl:1091
```
edit - Should have added that all tests pass and I'm on latest master of NCDatasets.

```julia
julia> Pkg.test("NCDatasets")
INFO: Testing NCDatasets
NetCDF library: /home/proy/.julia/v0.6/Conda/deps/usr/lib/libnetcdf.so
NetCDF version: 4.5.0 of Nov  8 2017 13:35:26 $
Test Summary: | Pass  Total
NCDatasets    |  547    547
INFO: NCDatasets tests passed

julia> Pkg.status("NCDatasets")
 - NCDatasets                    0.0.10+            master
```

**Comment by Alexander-Barth** [comment]
Thanks for the detailed info!
I think there are at least two issues here.
A easy to fix issue, with a null terminating character in the units attribute  is now fixed in master.

Another problem is the fact that NCDatasets currently supports only the  the standard (gregorian) calendar or Chronological Julian Date ( calendar is "julian" and and the time origin is -4713-01-01 ). The master version produces now a better error message.

The CF convention support several additional ("made up") calendars which  cannot be mapped to Julia's `DateTime`.  For instance `DateTime(2001,2,29)` is an error in Julia but it is a valid date in the all_leap calendar. For the noleap calendar, should we map all dates to the regular calendar ? If we do this, a model with a daily output in the noleap calendar, would sometimes have an time increment of 2 days according to the Julia's `DateTime`. Would this be an issue?

I had a quick check in Python's netCDF4, library and it seems that their `num2date` function would create custom objects (`netcdftime._netcdftime.DatetimeAllLeap` and `netcdftime._netcdftime.DatetimeNoLeap`)
in this case.

Note, in the meantime you can bypass the DateTime convertion by using e.g.:

```julia
ds = NCDatasets.Dataset("/home/abarth/Downloads/tasmax_day_IPSL-CM5A-MR_historical_r2i1p1_20000101-20051231.nc")
nctime = ds["time"].var[:]
```
Suggestions on what would be a good solution for these calendars would be welcome!

**Comment by Balinus** [comment]
Yeah, I've also looked at what python did and I wasn't interested to create new date type. 

Currently in ClimateTools.jl I deal with 2 main types of calendar. 

"noleap" || "365_day" 
and 
"gregorian" || "standard" || "proleptic_gregorian"

with some basic hacks. 

https://github.com/Balinus/ClimateTools.jl/blob/db418f765779569008f75b070d9a129da258f712/src/extract.jl#L558

In my view, having an increment of 2 days is not an issue.

Cheers!

**Comment by Alexander-Barth** [comment]
Can you try the CFDateTime branch? I tried to implement a similar approach than python's netCDF4 and cftime package which seem to me the safest approach.

It implements these new date types type:
```
      DateTimeStandard,
      DateTimeJulian,
      DateTimePGregorian,
      DateTimeAllLeap,
      DateTimeNoLeap,
      DateTime360
```

You can access, year, month, ... with the usual functions Dates.year, Dates.month, ...
Basic arithmetic (+ and -) and comparison (==, <, > , ≤  and ≥) is also supported.

For the "real" calendars types (standard/gregorian, proleptic gregorian, julian), there is an automatic conversion to the native julia's DateTime (which is by the way proleptic gregorian in Julia). This conversion rule is automatically applied when reading NetCDF files. The conversion preserve the duration (i.e. the difference) between two different dates.

The other calendars cannot be converted to julia's DateTime, but they can be reinterpreted (except some leap days) as a DateTime with the function `reinterprete`, e.g. 

```julia
dt = reinterpret(DateTime, DateTimeNoLeap(1900,2,28))
# returns DateTime(1900,2,28)
```
However, this is not done automatically (as the duration between different dates cannot be preserved).
When reading the time vector of your example data (calendar = "noleap"), you should thus get a special type `DateTimeNoLeap`. 

Does it work for you?

Thanks!

**Comment by Balinus** [comment]
Looks awesome! Will try asap.

**Comment by Balinus** [comment]
I can confirm that the branch works as expected. Awesome work!

Is there an unified type for these new types? The new calendar seems to share the `NCDatasets.AbstractCFDateTime` but not the typical calendar (e.g. `proleptic_gregorian` is not a subtype of `NCDatasets.AbstractCFDateTime`). Just wondering the best way to integrate your work in my functions.

Thanks again, I consider this work as a major milestone!

**Comment by Balinus** [comment]
Another question : is there a way to go back from the converted calendar (e.g. `DateTimeAllLeap`) to the original time vectors (i.e. `days since 0001-01-01`) ?

**Comment by Alexander-Barth** [comment]
Great! Thanks for you feed-back!
Indeed, AbstractCFDateTime is the base class, these subtypes should be defined:

```julia
subtypes(NCDatasets.AbstractCFDateTime)
6-element Array{Union{DataType, UnionAll},1}:
 NCDatasets.DateTime360       
 NCDatasets.DateTimeAllLeap   
 NCDatasets.DateTimeJulian    
 NCDatasets.DateTimeNoLeap    
 NCDatasets.DateTimePGregorian
 NCDatasets.DateTimeStandard  
```

But I think I should probably rename the type names to match the ones of cftime, e.g. DateTimePGregorian =>   DateTimeProlepticGregorian

The reverse operation is timeencode, for example:

```julia
timeencode([DateTimeJulian(2010,10,29,9,0,0)],"days since -4713-01-01T00:00:00","julian")
```
The function is defined as:
```
timeencode(data,units,calendar = "standard")
```
Maybe  date2num/num2date (from cftime.py) are better function names.

**Comment by Balinus** [comment]
Great, I do obtain the same subtype with `subtypes(NCDatasets.AbstractCFDateTime)`.

The problem I see is when my file contains Proleptic Gregorian. It returns a regular Array containing `DateTime`, which in turn is not a subtype of `AbstractCFDateTime`. I'm wondering if all possible calendar should share the same abstract super-type (?).

```julia
julia> ds4["time"]
time (2192)
  Datatype:    Float64
  Dimensions:  time
  Attributes:
   bounds               = time_bnds
   units                = days since 0001-01-01
   calendar             = proleptic_gregorian
   axis                 = T
   long_name            = time
   standard_name        = time

julia> timevec=ds4["time"][:]
2192-element DataArrays.DataArray{DateTime,1}:
 2000-01-01T12:00:00
 2000-01-02T12:00:00
 2000-01-03T12:00:00
 2000-01-04T12:00:00
 2000-01-05T12:00:00
 2000-01-06T12:00:00
 2000-01-07T12:00:00
 2000-01-08T12:00:00
 2000-01-09T12:00:00
 2000-01-10T12:00:00
 ⋮                  
 2005-12-23T12:00:00
 2005-12-24T12:00:00
 2005-12-25T12:00:00
 2005-12-26T12:00:00
 2005-12-27T12:00:00
 2005-12-28T12:00:00
 2005-12-29T12:00:00
 2005-12-30T12:00:00
 2005-12-31T12:00:00

julia> typeof(timevec[1]) <: NCDatasets.AbstractCFDateTime
false
```
Here's the file with the proleptic calendar.
https://www.dropbox.com/s/qcvpjtr7kyqjded/tasmax_day_ACCESS1-3_historical_r1i1p1_20000101-20051231.nc?dl=0

**Comment by Alexander-Barth** [comment]
AbstractCFDateTime is now a subtype of Dates.TimeType (as is DateTime).
This should work now:

```julia
typeof(timevec[1]) <: Dates.TimeType 
```
If you want to be more specific you can also use:
```julia
typeof(timevec[1]) <: Union{DateTime,NCDatasets.AbstractCFDateTime} 
```

Let me know if this solves the problem for you.

Note that I also change the type names:
```
 NCDatasets.DateTime360       
 NCDatasets.DateTimeAllLeap   
 NCDatasets.DateTimeJulian    
 NCDatasets.DateTimeNoLeap    
 NCDatasets.DateTimePGregorian
 NCDatasets.DateTimeStandard  
```

**Comment by Balinus** [comment]
Nice! I'll migrate those time conversion on your package asap. 😀

I'll test it out in the coming week. Thanks again fo this work!

**Comment by Balinus** [comment]
Is there a quick-way to broadcast the `timeencode` function? 

e.g.:

```julia
julia> using NCDatasets
julia> ds=Dataset(cmip5file2)
julia> ds["time"]
time (2190)
  Datatype:    Float64
  Dimensions:  time
  Attributes:
   bounds               = time_bnds
   units                = days since 1850-01-01 00:00:00
   calendar             = noleap
   axis                 = T
   long_name            = time
   standard_name        = time

timevec =ds["time"][:]
2190-element DataArrays.DataArray{NCDatasets.DateTimeNoLeap,1}:
 NCDatasets.DateTimeNoLeap(2000-01-01T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-02T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-03T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-04T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-05T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-06T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-07T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-08T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-09T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-10T12:00:00)
 ⋮                                             
 NCDatasets.DateTimeNoLeap(2005-12-23T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-24T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-25T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-26T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-27T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-28T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-29T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-30T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-31T12:00:00)

julia> timeencode([timevec[10]], ds["time"].attrib["units"], ds["time"].attrib["calendar"])
1-element Array{Float64,1}:
 54759.5

julia> t=timeencode.([timevec[:]], ds["time"].attrib["units"], ds["time"].attrib["calendar"])
1-element Array{DataArrays.DataArray{NCDatasets.DateTimeNoLeap,1},1}:
 Union{Missings.Missing, NCDatasets.DateTimeNoLeap}[NCDatasets.DateTimeNoLeap(2000-01-01T12:00:00), NCDatasets.DateTimeNoLeap(2000-01-02T12:00:00), NCDatasets.DateTimeNoLeap(2000-01-03T12:00:00), NCDatasets.DateTimeNoLeap(2000-01-04T12:00:00), NCDatasets.DateTimeNoLeap(2000-01-05T12:00:00), NCDatasets.DateTimeNoLeap(2000-01-06T12:00:00), NCDatasets.DateTimeNoLeap(2000-01-07T12:00:00), NCDatasets.DateTimeNoLeap(2000-01-08T12:00:00), NCDatasets.DateTimeNoLeap(2000-01-09T12:00:00), NCDatasets.DateTimeNoLeap(2000-01-10T12:00:00)  …  NCDatasets.DateTimeNoLeap(2005-12-22T12:00:00), NCDatasets.DateTimeNoLeap(2005-12-23T12:00:00), NCDatasets.DateTimeNoLeap(2005-12-24T12:00:00), NCDatasets.DateTimeNoLeap(2005-12-25T12:00:00), NCDatasets.DateTimeNoLeap(2005-12-26T12:00:00), NCDatasets.DateTimeNoLeap(2005-12-27T12:00:00), NCDatasets.DateTimeNoLeap(2005-12-28T12:00:00), NCDatasets.DateTimeNoLeap(2005-12-29T12:00:00), NCDatasets.DateTimeNoLeap(2005-12-30T12:00:00), NCDatasets.DateTimeNoLeap(2005-12-31T12:00:00)]

julia> t[1]
2190-element DataArrays.DataArray{NCDatasets.DateTimeNoLeap,1}:
 NCDatasets.DateTimeNoLeap(2000-01-01T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-02T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-03T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-04T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-05T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-06T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-07T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-08T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-09T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-10T12:00:00)
 ⋮                                             
 NCDatasets.DateTimeNoLeap(2005-12-23T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-24T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-25T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-26T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-27T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-28T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-29T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-30T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-31T12:00:00)
```

**Comment by Alexander-Barth** [comment]
No broadcasting, but this should work on a vector (or array):

```julia
 t = timeencode(timevec, ds["time"].attrib["units"], ds["time"].attrib["calendar"])
```

There is the function definition of `timeencode`:

https://github.com/Alexander-Barth/NCDatasets.jl/blob/CFDateTime/src/time.jl#L531

**Comment by Balinus** [comment]
It does work but does not return the expected values.

```julia
julia> t = timeencode(timevec, ds["time"].attrib["units"], ds["time"].attrib["calendar"])
2190-element DataArrays.DataArray{NCDatasets.DateTimeNoLeap,1}:
 NCDatasets.DateTimeNoLeap(2000-01-01T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-02T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-03T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-04T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-05T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-06T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-07T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-08T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-09T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-10T12:00:00)
 ⋮                                             
 NCDatasets.DateTimeNoLeap(2005-12-23T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-24T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-25T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-26T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-27T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-28T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-29T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-30T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-31T12:00:00)

julia> t[1]
NCDatasets.DateTimeNoLeap(2000-01-01T12:00:00)
```

While doing a loop over timevec seems to be correct.

```julia
julia> t = Array{Any}(length(timevec))
julia> for i=1:length(timevec)
       t[i] = timeencode([timevec[i]], ds["time"].attrib["units"], ds["time"].attrib["calendar"])
       end

julia> t
2190-element Array{Any,1}:
 [54750.5]
 [54751.5]
 [54752.5]
 [54753.5]
 [54754.5]
 [54755.5]
 [54756.5]
 [54757.5]
 [54758.5]
 [54759.5]
 ⋮        
 [56931.5]
 [56932.5]
 [56933.5]
 [56934.5]
 [56935.5]
 [56936.5]
 [56937.5]
 [56938.5]
 [56939.5]
```

**Comment by Alexander-Barth** [comment]
Is it possible for you to make a small reproducible (independently from a netcdf file) test case out of this issue?

**Comment by Balinus** [comment]
While trying to reproduce the error, I've found what was the problem (I wasn't able to reproduce it!).

The type of `timevec` is a DataArray{NCDatasets.DateTimeNoLeap}. When converting to a regular Array, the expected results is given. For example, with the original data (independent test case below).

```julia
julia> typeof(timevec)
DataArrays.DataArray{NCDatasets.DateTimeNoLeap,1}

julia> timeencode(timevec, ds["time"].attrib["units"], ds["time"].attrib["calendar"])
2190-element DataArrays.DataArray{NCDatasets.DateTimeNoLeap,1}:
 NCDatasets.DateTimeNoLeap(2000-01-01T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-02T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-03T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-04T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-05T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-06T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-07T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-08T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-09T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-10T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-11T12:00:00)
 ⋮                                             
 NCDatasets.DateTimeNoLeap(2005-12-21T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-22T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-23T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-24T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-25T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-26T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-27T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-28T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-29T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-30T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-31T12:00:00)

julia> timevec2=Array{NCDatasets.DateTimeNoLeap, 1}(timevec)
2190-element Array{NCDatasets.DateTimeNoLeap,1}:
 NCDatasets.DateTimeNoLeap(2000-01-01T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-02T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-03T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-04T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-05T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-06T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-07T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-08T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-09T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-10T12:00:00)
 NCDatasets.DateTimeNoLeap(2000-01-11T12:00:00)
 ⋮                                             
 NCDatasets.DateTimeNoLeap(2005-12-21T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-22T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-23T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-24T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-25T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-26T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-27T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-28T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-29T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-30T12:00:00)
 NCDatasets.DateTimeNoLeap(2005-12-31T12:00:00)

julia> timeencode(timevec2, ds["time"].attrib["units"], ds["time"].attrib["calendar"])
2190-element Array{Float64,1}:
 54750.5
 54751.5
 54752.5
 54753.5
 54754.5
 54755.5
 54756.5
 54757.5
 54758.5
 54759.5
 54760.5
     ⋮  
 56929.5
 56930.5
 56931.5
 56932.5
 56933.5
 56934.5
 56935.5
 56936.5
 56937.5
 56938.5
 56939.5

```

Here's a MWE

```julia
# MWE with an Array
units = "days since 1850-01-01 00:00:00"
calendar = "noleap"
data_orig = Array([54750.5, 54751.5, 54752.5])

# Decoding
datacal = timedecode(data_orig, units, calendar)
# Reencoding
data_orig_back = timeencode(datacal, units, calendar)

# DataArray
data_orig = DataArrays.DataArray([54750.5, 54751.5, 54752.5])
# Decoding
datacal = timedecode(data_orig, units, calendar)
# Reencoding
data_orig_back = timeencode(datacal, units, calendar)
```

**Comment by Alexander-Barth** [comment]
Thanks a lot for this MWE! It was really helpful to identify the issue (a too narrow type definition for timeencode).
It is fixed now in CFDateTime and your MWE is part of the tests to avoid any regressions.

Thanks a again for your thoroughly testing this package!

**Comment by Balinus** [comment]
Nice! No problem, I plan to use it in the end! 😀

**Comment by Alexander-Barth** [comment]
Just to let you know that I now released version v0.1.0 with the CFDateTime branch merged.
I propose to close this issue if this is OK for your.

**Comment by Balinus** [comment]
Yep, thanks for your work!
